benchmarks/float8/utils.py

```diff
@@ -350,3 +336,326 @@ def get_gpu_kernel_gemm_time_s(f, *args, **kwargs):
+        ffn_dim_multiplier (Optional[float]): Custom multiplier for hidden dimension. Defaults to None.
+
+    Attributes:
+        w1 (Linear): Linear transformation for the first layer.
+        w2 (Linear): Linear transformation for the second layer.
+        w3 (Linear): Linear transformation for the third layer.
+
+    """
+
+    def __init__(
+        self,
+        dim: int,
+        hidden_dim: int,
+        multiple_of: int,
+        ffn_dim_multiplier: Optional[float],
+    ):
+        super().__init__()
+        hidden_dim = int(2 * hidden_dim / 3)
+        # custom dim factor multiplier
+        if ffn_dim_multiplier is not None:
+            hidden_dim = int(ffn_dim_multiplier * hidden_dim)
+        hidden_dim = multiple_of * ((hidden_dim + multiple_of - 1) // multiple_of)
+
+        self.w1 = nn.Linear(dim, hidden_dim, bias=False)
+        self.w2 = nn.Linear(hidden_dim, dim, bias=False)
+        self.w3 = nn.Linear(dim, hidden_dim, bias=False)
+
+    def forward(self, x):
+        return self.w2(F.silu(self.w1(x)) * self.w3(x))
+
+    def init_weights(self, init_std: float):
+        nn.init.trunc_normal_(self.w1.weight, mean=0.0, std=0.02)
+        for linear in (self.w2, self.w3):
+            nn.init.trunc_normal_(linear.weight, mean=0.0, std=init_std)
+
+
+class NormFFNResidualNorm(nn.Module):
+    """
+    A fragment representing the end of TransformerBlock n and the start
+    of TransformerBlock n + 1, intended to include the fusions relevant
+    to float8 gemms in the FFN module in forward and backward.
+    """
+
+    def __init__(self, dim, hidden_dim, multiple_of, ffn_dim_multiplier):
+        super().__init__()
+        self.ffn_norm = RMSNorm(dim)
+        self.ffn = FeedForward(dim, hidden_dim, multiple_of, ffn_dim_multiplier)
+        self.attn_norm = RMSNorm(dim)
+
+    def forward(self, h):
+        # end of transformer block n
+        x = self.ffn_norm(h)
+        x = self.ffn(x)
+        x = h + x
+        # start of transformer block n + 1
+        x = self.attn_norm(x)
+        return x
+
+#
+# The following model definitions intend to capture all the 
+# relevant (prev_op -> linear -> [next_op, ...]) subgraphs from the LLaMa model
+# family.  We use https://github.com/pytorch/torchtitan/tree/main/torchtitan/models/llama
+# as the reference code structure, and manually define all the subgraphs
+# around linear layers.
+#
+# The goal here is to set up microbenchmarks on the fusion of float8 
+# scaling/casting that are relevant to transformer models, and break
+# them up in such a way that we can easily reason about fusions regarding
+# each linear layer without having to dig through large traces / long log files.
+# There is some duplication here (attn_norm and ffn_norm appear in multiple
+# places), and that's ok.
+#
+# Simplifying assumptions (may be relaxed in the future):
+# 1. don't model fusion with rotary embeddings and repeat-transpose
+# 2. don't model fusion with SDPA
```

Comment:
cc @drisspg , curious on your thoughts on how future-proof this assumption is

Reply:
mmmmm I think this is okay, the haddamard transform from FA3 is something we would expect to be fused but I think we can just leave a TODO until we have low-p fused attention training support